ExportToDatabase module - AssertionError: Feature ExportToDb_Images when running headlessly

We are using cellprofiler based workflow to process our high-content screen output. Workflow runs without issues when exporting to spreadsheet within an unique `plateid_wellid/` output directory. While exporting to database using MySQL server, I am able to create a database using ExportToDatabase module when I initialize run for the first well, say `plate1_A01`. However, when I run a second well, say `plate1_A02` (either in parallel or after finishing `plate1_A01`), I get an error related to following:

```
  File "/projects/verhaak-lab/amins/hpcenv/mambaforge/envs/cp2/lib/python3.8/site-packages/cellprofiler_core/utilities/hdf5_dict.py", line 435, in __getitem__
    assert feature_exists, "Feature {} for {} does not exist".format(
AssertionError: Feature ExportToDb_Images for Experiment does not exist
Wed Dec  7 15:43:02 2022: Image # 1, module ExportToDatabase # 26: CPU_time = 0.20 secs, Wall_time = 0.33 secs
```

>Complete log, including database schema, workflow.cppipe file in dropbox toyset link below.

I believe this issue is related to #4471 but I am unable to resolve it using an updated cellprofiler version 4.2.4

Related to database export, I also have not figured out a way to populate `Per Image` data such that running multiple wells across multiple plates in parallel does not overwrite `Per Image` entries (currently set at field level, 9 fields or rows): For now, I cannot find  a way to modify `ImageNumber` column such that it inherits `plate_well` specific information (either in the same table or linked with key constraint from an another table). Ideally, we like to leverage MySQL to store data coming out of 55K wells across 144 384-well plates in parallel (each HPC job processes a single well). I read related forum posts, https://forum.image.sc/search?q=ExportToDatabase%20overwrite but finding it difficult to resolve these two issues.

**To Reproduce**
Steps to reproduce the behavior:

Please download toyset: https://go.sbamin.com/aedM (`cpdebug`) with a directory structure for `v20221207/`
 
```sh
## workflow
workflow="<path to>/synhcs_workflow_seg_v7.2_gitrm.cppipe"

## input data
master_wellid_path="<path to>/maxp/output/SCBHCS000767_T1_C01"

## output dir
cellprofiler_outdir="${master_wellid_path}"/segrun
mkdir -p "${cellprofiler_outdir}"

## Run cellprofiler
## Database connection info and plugin path needs to be updated
## toyset is not using plugins (stardist and cellpose), so workflow may run without it.
cellprofiler \
--run-headless \
--run \
--pipeline "${workflow}" \
--output-directory "${cellprofiler_outdir}" \
--image-directory "${master_wellid_path}" \
--log-level 20 \
--plugins-directory /projects/verhaak-lab/amins/hpcenv/opt/modules/apps/cellprofiler/CellProfiler-plugins
```

Above code should run for the first well, `SCBHCS000767_T1_C01` or `SCBHCS000769_T2_G18` but will fail with an error as above while running a subsequent well.

**Expected behavior**
* Workflow should run for all wells, ideally while submitting all 384 well jobs in parallel, and should aggregate data in MySQL
* MySQL database `Per_Image` table should not overwrite Image level data, and rather retain plate and well specific `ImageNumber` and corresponding output.

**Screenshots**
Given under toyset folder, `v20221207/log/mysql_schema/`

**Desktop (please complete the following information):**
 - OS: Os Monterey 12.6
 - CellProfiler Version 4.2.4
 - pip install using conda recipe (along with GPU-based CellPose and StarDist packages in the same conda env)

Thanks!

Hi Beth,

I added batch file using CreateBatchFiles module at the very end of the workflow and created a batch file. Rest of workflow is identical to the one I put in toyset above.
![Screen Shot 2022-12-12 at 2 30 34 PM](https://user-images.githubusercontent.com/12156027/207136839-eaa666a7-711c-4efc-aeae-26a9b1bdfe8a.png)

I copied that batch file to out HPC and kept it in an output folder that was mapped as above to my macbook base directory. I have identical folder structure on our HPC at `/fastscratch/amins/hcs/` to that of macbbook, `/Users/amins/scratch/cell_profiler/toyset` folder.

Now when I run following command on HPC, I get the same error, *AssertionError: Feature ExportToDb_Images for Experiment does not exist*. This error stayed the same regardless of if I am using a pre-existing db schema (created while running AnalyzeImages with creating batch file on macbook) or starting with a no database at all on MySQL server. Error also remained same  if I use `-g Metadata_Plate=SCBHCS000767_T1,Metadata_Well=C01` or `-f 1 -l 18` option.

```sh
cellprofiler \
-c -r -p /fastscratch/amins/hcs/output/wip/synhcs_workflow_seg_v7.2_gitrm.h5 \
-g Metadata_Plate=SCBHCS000767_T1,Metadata_Well=C01 \
--plugins-directory /projects/verhaak-lab/amins/hpcenv/opt/modules/apps/cellprofiler/CellProfiler-plugins
```

Error log:

```
> -c -r -p /fastscratch/amins/hcs/output/wip/synhcs_workflow_seg_v7.2_gitrm.h5 \> -g Metadata_Plate=SCBHCS000767_T1,Metadata_Well=C01 \
> --plugins-directory /projects/verhaak-lab/amins/hpcenv/opt/modules/apps/cellprofiler/CellProfiler-plugins
Experiment-wide values for mean threshold, etc calculated by MeasureImageQuality may be incorrect if the run is split into subsets of images.
Set MySQL transaction isolation to "READ COMMITTED": 0Batch file default input directory "/Users/amins", does not exist
Times reported are CPU and Wall-clock times for each module
...
Mon Dec 12 14:34:05 2022: Image # 1, module MeasureImageIntensity # 22: CPU_time = 0.26 secs, Wall_time = 0.23 secs
saveimages.py (791): /fastscratch/amins/hcs/output/wip/seg/SCBHCS000767_T1_C01_F01_maxp_rgb.tiff is a low contrast image
Mon Dec 12 14:34:05 2022: Image # 1, module SaveImages # 23: CPU_time = 0.06 secs, Wall_time = 0.09 secs
saveimages.py (791): /fastscratch/amins/hcs/output/wip/seg/SCBHCS000767_T1_C01_F01_overlay_segs.tiff is a low contrast image
Mon Dec 12 14:34:05 2022: Image # 1, module SaveImages # 24: CPU_time = 0.06 secs, Wall_time = 0.07 secs
Mon Dec 12 14:34:05 2022: Image # 1, module ExportToSpreadsheet # 25: CPU_time = 0.00 secs, Wall_time = 0.00 secs
pipeline_exception
Error detected during run of module ExportToDatabase
Traceback (most recent call last):
  File "/projects/verhaak-lab/amins/hpcenv/mambaforge/envs/cp2/lib/python3.8/site-packages/cellprofiler_core/pipeline/_pipeline.py", line 1008, in run_with_yield
    self.run_module(module, workspace)
  File "/projects/verhaak-lab/amins/hpcenv/mambaforge/envs/cp2/lib/python3.8/site-packages/cellprofiler_core/pipeline/_pipeline.py", line 1341, in run_module
    module.run(workspace)
  File "/projects/verhaak-lab/amins/hpcenv/mambaforge/envs/cp2/lib/python3.8/site-packages/cellprofiler/modules/exporttodatabase.py", line 2585, in run
    self.record_image_channels(workspace)
  File "/projects/verhaak-lab/amins/hpcenv/mambaforge/envs/cp2/lib/python3.8/site-packages/cellprofiler/modules/exporttodatabase.py", line 4794, in record_image_channels
    image_list = workspace.measurements.get_experiment_measurement("ExportToDb_Images")
  File "/projects/verhaak-lab/amins/hpcenv/mambaforge/envs/cp2/lib/python3.8/site-packages/cellprofiler_core/measurement/_measurements.py", line 973, in get_experiment_measurement
    result = self.get_measurement(EXPERIMENT, feature_name)
  File "/projects/verhaak-lab/amins/hpcenv/mambaforge/envs/cp2/lib/python3.8/site-packages/cellprofiler_core/measurement/_measurements.py", line 848, in get_measurement
    result = self.hdf5_dict[EXPERIMENT, feature_name, 0]
  File "/projects/verhaak-lab/amins/hpcenv/mambaforge/envs/cp2/lib/python3.8/site-packages/cellprofiler_core/utilities/hdf5_dict.py", line 430, in __getitem__
    result = self[object_name, feature_name, [num_idx]]
  File "/projects/verhaak-lab/amins/hpcenv/mambaforge/envs/cp2/lib/python3.8/site-packages/cellprofiler_core/utilities/hdf5_dict.py", line 435, in __getitem__
    assert feature_exists, "Feature {} for {} does not exist".format(
AssertionError: Feature ExportToDb_Images for Experiment does not exist
Mon Dec 12 14:34:05 2022: Image # 1, module ExportToDatabase # 26: CPU_time = 0.12 secs, Wall_time = 0.13 secs
So sorry. CellProfiler failed to remove the temporary file, /tmp/Cpmeasurementsi5feq24y.hdf5 and there it sits on your disk now.
```

--- a/cellprofiler/modules/exporttodatabase.py
+++ b/cellprofiler/modules/exporttodatabase.py
@@ -4210,17 +4210,7 @@ CREATE TABLE %s (
         else:
             object_names = [None]
 
-        default_image_names = []
-        # Find all images that have FileName and PathName
-        image_features = [
-            c[1]
-            for c in workspace.pipeline.get_measurement_columns()
-            if c[0] == "Image"
-        ]
-        for feature in image_features:
-            match = re.match("^%s_(.+)$" % C_FILE_NAME, feature)
-            if match:
-                default_image_names.append(match.groups()[0])
+        default_image_names = self.get_image_list(workspace)
 
         if not self.properties_export_all_image_defaults:
             # Extract the user-specified images
@@ -4793,6 +4783,9 @@ process_3D = {process_3D}
     def record_image_channels(self, workspace):
         # We only have access to the image details during the run itself.
         # Fetch out the images we want in the properties file and log their channel counts.
+        if not workspace.measurements.hdf5_dict.has_feature("Experiment", "ExportToDb_Images"):
+            image_list = self.get_image_list(workspace)
+            workspace.measurements.add_experiment_measurement("ExportToDb_Images", image_list)
         image_list = workspace.measurements.get_experiment_measurement("ExportToDb_Images")
         channel_list = []
         if isinstance(image_list, str):
@@ -4953,6 +4946,20 @@ CP version : %d\n""" % int(
         object_name - name of object or "Image", "Object" or "Well"
         """
         return self.get_table_prefix() + "Per_" + object_name
+
+    def get_image_list(self,workspace):
+        default_image_names = []
+        # Find all images that have FileName and PathName
+        image_features = [
+            c[1]
+            for c in workspace.pipeline.get_measurement_columns()
+            if c[0] == "Image"
+        ]
+        for feature in image_features:
+            match = re.match("^%s_(.+)$" % C_FILE_NAME, feature)
+            if match:
+                default_image_names.append(match.groups()[0])
+        return default_image_names
 
     def get_pipeline_measurement_columns(
         self, pipeline, image_set_list, remove_postgroup_key=False

